src/anchor-navigation/scroll-spy.tsx

```diff
@@ -0,0 +1,77 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+const isBrowser = typeof window !== 'undefined';
+
+export default function useScrollSpy({
+  hrefs,
+  scrollSpyOffset,
+  disableScrollSpy = false,
+}: {
+  hrefs: string[];
+  scrollSpyOffset: number;
+  disableScrollSpy?: boolean;
+}): [string | undefined, React.Dispatch<React.SetStateAction<string | undefined>>] {
+  const [currentHref, setCurrentHref] = useState<string>();
+
+  const lastAnchorElementExists = useMemo(() => !!document.getElementById(hrefs[hrefs.length - 1]?.slice(1)), [hrefs]);
+
+  // Get the bounding rectangle of an element by href
+  const getRectByHref = useCallback(href => {
+    const element = document.getElementById(href.slice(1));
+    return element?.getBoundingClientRect();
+  }, []);
+
+  // Check if we're scrolled to the bottom of the page
+  const isPageBottom = useCallback(() => {
+    return lastAnchorElementExists && window.scrollY >= Math.floor(document.body.scrollHeight - window.innerHeight);
+  }, [lastAnchorElementExists]);
+
+  // Find the href for which the element is within the viewport
+  const findHrefInView = useCallback(() => {
+    console.log('Here');
+    return hrefs.find(href => {
+      const rect = getRectByHref(href);
+      return rect && rect.bottom <= window.innerHeight && rect.bottom - scrollSpyOffset >= 0;
+    });
+  }, [getRectByHref, scrollSpyOffset, hrefs]);
+
+  // Find the last href where its element is above or within the viewport
+  const findLastHrefInView = useCallback(() => {
+    return hrefs
+      .slice()
+      .reverse()
+      .find(href => {
+        const rect = getRectByHref(href);
+        return rect && rect.bottom <= window.innerHeight;
+      });
+  }, [getRectByHref, hrefs]);
+
+  // Scroll event handler
+  const handleScroll = useCallback(() => {
+    if (disableScrollSpy || !isBrowser) {
+      return;
+    }
+
+    const scrollY = window.scrollY;
+
+    if (isPageBottom()) {
+      setCurrentHref(hrefs[hrefs.length - 1]);
+    } else {
+      setCurrentHref(findHrefInView() || (scrollY > 0 ? findLastHrefInView() : undefined));
+    }
+  }, [disableScrollSpy, isPageBottom, findHrefInView, findLastHrefInView, hrefs]);
+
+  useEffect(() => {
+    if (isBrowser) {
+      handleScroll();
+      window.addEventListener('scroll', handleScroll, { passive: true });
```

Comment:
Did you consider using the Intersection Observer API instead? https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

Reply:
I did consider it initially. In fact, I implemented it (https://github.com/cloudscape-design/components/pull/1396/commits/00a711b3bcaf3421f1e6fdcc1657e3c503ca1422) but ran into some issues that made the code more complex than I would've liked...

Reply:
There were some edge cases (scrolled all the way to the bottom, in which the element didn't intersect the IO), lack of granularity, IO sometimes not being triggered (when being linked to a specific section of the page or quickly scrolling on the trackpad)